core/src/main/java/org/elasticsearch/index/snapshots/IndexShardSnapshotStatus.java

```diff
@@ -47,119 +50,94 @@
+        } else {
+            throw new IllegalStateException("Unable to move the shard snapshot status to [FINALIZE]: " +
+                "expecting [STARTED] but got [" + stage.get() + "]");
+        }
+        return asCopy();
     }
 
-    /**
-     * Sets snapshot start time
-     *
-     * @param startTime snapshot start time
-     */
-    public void startTime(long startTime) {
-        this.startTime = startTime;
+    public synchronized Copy moveToDone(final long endTime) {
+        if (stage.compareAndSet(Stage.FINALIZE, Stage.DONE)) {
+            this.totalTime = Math.max(0L, endTime - startTime);
+        } else {
+            throw new IllegalStateException("Unable to move the shard snapshot status to [DONE]: " +
+                "expecting [FINALIZE] but got [" + stage.get() + "]");
+        }
+        return asCopy();
     }
 
-    /**
-     * Returns snapshot processing time
-     *
-     * @return processing time
-     */
-    public long time() {
-        return this.time;
+    public synchronized Copy moveToAborted(final String failure) {
+        if (stage.getAndSet(Stage.ABORTED) != Stage.ABORTED) {
+            this.failure = failure;
+        }
+        return asCopy();
     }
 
-    /**
-     * Sets snapshot processing time
-     *
-     * @param time snapshot processing time
-     */
-    public void time(long time) {
-        this.time = time;
+    public synchronized Copy abortIfNotCompleted(final String failure) {
+        if (stage.compareAndSet(Stage.INIT, Stage.ABORTED) || stage.compareAndSet(Stage.STARTED, Stage.ABORTED)) {
+            this.failure = failure;
+        }
+        return asCopy();
     }
 
-    /**
-     * Returns true if snapshot process was aborted
-     *
-     * @return true if snapshot process was aborted
-     */
-    public boolean aborted() {
-        return this.aborted;
+    public synchronized void moveToFailed(final long endTime, final String failure) {
+        if (stage.getAndSet(Stage.FAILURE) != Stage.FAILURE) {
+            this.totalTime = Math.max(0L, endTime - startTime);
+            this.failure = failure;
+        }
     }
 
-    /**
-     * Marks snapshot as aborted
-     */
-    public void abort() {
-        this.aborted = true;
-    }
-
-    /**
-     * Sets files stats
-     *
-     * @param numberOfFiles number of files in this snapshot
-     * @param totalSize     total size of files in this snapshot
-     */
-    public void files(int numberOfFiles, long totalSize) {
-        this.numberOfFiles = numberOfFiles;
-        this.totalSize = totalSize;
-    }
-
-    /**
-     * Sets processed files stats
-     *
-     * @param numberOfFiles number of files in this snapshot
-     * @param totalSize     total size of files in this snapshot
-     */
-    public synchronized void processedFiles(int numberOfFiles, long totalSize) {
-        processedFiles = numberOfFiles;
-        processedSize = totalSize;
+    public void ensureNotAborted() {
```

Comment:
I would prefer to have the previous boolean `isAborted` method. Call sites currently catch this exception and then throw another exception.

Reply:
Ok, I revert the changes.